File: daily_problems/2026/04/0417/personal_submission/cf105501g_yawn_sean.cpp
#include <bits/stdc++.h>
// #pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/all"

using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

// Let's unordered_set a bit and flow!
// I came, I divided, I conquered!
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int n;
	cin >> n;

	vector<int> graph(n);
	for (auto &x: graph) {
		cin >> x;
		x --;
	}

	vector<vector<int>> rev_graph(n);
	for (int i = 0; i < n; i ++)
		rev_graph[graph[i]].emplace_back(i);

	vector<int> indeg(n, 0);
	for (int x : graph)
		indeg[x] ++;

	vector<int> stk;
	for (int i = 0; i < n; i ++)
		if (indeg[i] == 0) stk.emplace_back(i);

	while (!stk.empty()) {
		int u = stk.back(); stk.pop_back();
		indeg[graph[u]] --;
		if (indeg[graph[u]] == 0)
			stk.emplace_back(graph[u]);
	}

	vector<int> vis(n, 0);
	int ans = 0;

	auto dfs = [&] (auto &self, int u) -> pair<int, int> {
		int v0 = 0, v1 = 1;
		for (auto &v: rev_graph[u]) {
			if (!vis[v]) {
				auto [x0, x1] = self(self, v);
				v0 += max(x0, x1);
				v1 += x0;
			}
		}
		return {v0, v1};
	};

	for (int i = 0; i < n; i ++) {
		if (indeg[i] && !vis[i]) {
			vector<int> cycle = {i};
			vis[i] = 1;
			while (graph[cycle.back()] != i) {
				cycle.emplace_back(graph[cycle.back()]);
				vis[cycle.back()] = 1;
			}

			vector<int> v0s, v1s;

			for (auto &u: cycle) {
				auto [v0, v1] = dfs(dfs, u);
				v0s.emplace_back(v0);
				v1s.emplace_back(v1);
			}

			int k = v0s.size();
			int res = 0;

			int val0 = 0, val1 = 0;
			for (int j = 1; j < k; j ++) {
				int nv0 = max(val0, val1) + v0s[j];
				int nv1 = val0 + v1s[j];
				val0 = nv0; val1 = nv1;
			}
			res = max(res, v0s[0] + max(val0, val1));

			val0 = 0; val1 = 0;
			for (int j = 0; j < k - 1; j ++) {
				int nv0 = max(val0, val1) + v0s[j];
				int nv1 = val0 + v1s[j];
				val0 = nv0; val1 = nv1;
			}
			res = max(res, v0s[k - 1] + max(val0, val1));

			ans += res;
		}
	}

	cout << n - ans;

	return 0;
}

File: daily_problems/2026/04/0417/personal_submission/cf105501g_yawn_sean.py
# Submission link: https://codeforces.com/gym/105501/submission/371339432
def main():
    n = II()
    graph = LGMI()
    
    rev_graph = [[] for _ in range(n)]
    for i in range(n):
        rev_graph[graph[i]].append(i)
    
    indeg = [0] * n
    for x in graph:
        indeg[x] += 1
    
    stk = [i for i in range(n) if indeg[i] == 0]
    while stk:
        u = stk.pop()
        indeg[graph[u]] -= 1
        if indeg[graph[u]] == 0:
            stk.append(graph[u])
    
    vis = [0] * n
    
    dp0 = [0] * n
    dp1 = [0] * n
    
    ans = 0
    
    for i in range(n):
        if indeg[i] and not vis[i]:
            cycle = [i]
            vis[i] = 1
            
            while graph[cycle[-1]] != i:
                cycle.append(graph[cycle[-1]])
                vis[cycle[-1]] = 1
            
            v0s = []
            v1s = []
            
            for u in cycle:
                que = [u]
                for x in que:
                    for y in rev_graph[x]:
                        if not vis[y]:
                            vis[y] = 1
                            que.append(y)
                
                for x in reversed(que):
                    dp1[x] = 1
                    for y in rev_graph[x]:
                        if indeg[y] == 0:
                            dp0[x] += fmax(dp0[y], dp1[y])
                            dp1[x] += dp0[y]
    
                v0s.append(dp0[u])
                v1s.append(dp1[u])
            
            k = len(v0s)
            res = 0
            
            val0 = 0
            val1 = 0
            
            for i in range(1, k):
                val0, val1 = fmax(val0, val1) + v0s[i], val0 + v1s[i]
            
            res = fmax(res, v0s[0] + fmax(val0, val1))
    
            val0 = 0
            val1 = 0
            
            for i in range(k - 1):
                val0, val1 = fmax(val0, val1) + v0s[i], val0 + v1s[i]
            
            res = fmax(res, v0s[-1] + fmax(val0, val1))
            
            ans += res
    
    print(n - ans)